### PR TITLE
Add intermediate containers

### LINKED ISSUE #1221: Prevent rebuilding of all containers when changing package.json
I added a dep in package.json and wanted to install the new package. Sadly, `make serve` causes the container(s) to install basically from scratch (from `apt-get update` on). Is there a better way to install additional packages or add new frontend js dependencies?

### PR DESCRIPTION
**Pull request by di**
Changes to the working directory can force a rebuild of the containers after the `COPY . /app/` step. At the moment, this includes downloading a fair amount of dependencies, none of which need to happen after copying the Warehouse source into the container.

This PR moves the installation of these dependencies into a series of intermediate containers before finally performing the `COPY . /app/` step.

Fixes #1221.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -23,38 +23,49 @@ RUN set -x \
...
+
+COPY package.json /tmp/package.json
+
+# Install NPM dependencies
+RUN set -x \
     && npm install -g phantomjs-prebuilt gulp-cli \
+    && cd /tmp \
     && npm install \
+    && mkdir /app \
+    && cp -a /tmp/node_modules /app/
```
Comment (dstufft) [inline Dockerfile]: Huh, does this actually work?

File: Dockerfile [hunk 2]
```diff
@@ -23,38 +23,49 @@ RUN set -x \
...
     # Uncomment the below line if you're working on the PyPI theme, this is a
     # private repository due to the fact that other people's IP is contained
     # in it.
     # && pip install -c requirements/main.txt -r requirements/theme.txt \
     && find /usr/local -type f -name '*.pyc' -name '*.pyo' -delete \
-    && rm -rf ~/.cache/ \
+    && rm -rf ~/.cache/
+
+# Clean up build dependencies
+RUN set -x \
```
Comment (dstufft) [inline Dockerfile]: If we're not going to do this inline with when we install them, then there's not a lot of point to ever uninstalling them. It's just a trick to keep the image size smaller, but that won't happen if you don't install, use, uninstall all in the same ``RUN`` command.

Comment (di) [inline Dockerfile]: Good point. I'm forgetting now why I thought it should be a separate command... 🙂

File: Dockerfile (hunk 1 above)
Comment (di) [inline Dockerfile]: Indeed!

**Pull request merged**